File: LeetCode.com/75-SortColors.cs
//75. Sort Colors

public class Solution {
    public void SortColors(int[] nums) {
        int[] counts = new int[3];
        foreach (int x in nums)  {
            counts[x]++;
        }
        
        Console.WriteLine(String.Join(",", counts));
        
        int index = 0;
        
        for (int i = 0; i <= 2; i++)  {
            while (counts[i]-- > 0) {
                nums[index++] = i;
            }
        }
    }
}

File: LeetCode Java/Data Structure II/75-SortColors.java
/*
75. Sort Colors
Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.
We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.
You must solve this problem without using the library's sort function.
Example 1:
    Input: nums = [2,0,2,1,1,0]
    Output: [0,0,1,1,2,2]
*/

class Solution {
  /*
  Dutch National Flag problem solution.
  Time complexity : O(N) since it's one pass along the array of length N.
  Space complexity : O(1) since it's a constant space solution.
  */
  public void sortColors(int[] nums) {
    int p0 = 0;
    int curr = 0;
    int p2 = nums.length - 1;

    int tmp;
    while (curr <= p2) {
      if (nums[curr] == 0) {
        // swap p0-th and curr-th elements
        // i++ and j++
        tmp = nums[p0];
        nums[p0++] = nums[curr];
        nums[curr++] = tmp;
      }
      else if (nums[curr] == 2) {
        // swap k-th and curr-th elements
        // p2--
        tmp = nums[curr];
        nums[curr] = nums[p2];
        nums[p2--] = tmp;
      }
      else curr++;
    }
  }
}